container_opts.go

```diff
@@ -324,3 +326,37 @@ func WithSpec(s *oci.Spec, opts ...oci.SpecOpts) NewContainerOpts {
 func WithoutRefreshedMetadata(i *InfoConfig) {
 	i.Refresh = false
 }
+
+// WithCDI updates OCI spec with CDI content
+func WithCDI(s *oci.Spec, annotations map[string]string, cdiSpecDirs []string) NewContainerOpts {
+	return func(ctx context.Context, _ *Client, c *containers.Container) error {
+		// TODO: Once CRI is extended with native CDI support this will need to be updated...
+		_, cdiDevices, err := cdi.ParseAnnotations(annotations)
+		if err != nil {
+			return fmt.Errorf("failed to parse CDI device annotations: %w", err)
+		}
+		if cdiDevices == nil {
+			return nil
+		}
+
+		registry := cdi.GetRegistry(cdi.WithSpecDirs(cdiSpecDirs...))
+		if err = registry.Refresh(); err != nil {
+			// We don't consider registry refresh failure a fatal error.
+			// For instance, a dynamically generated invalid CDI Spec file for
+			// any particular vendor shouldn't prevent injection of devices of
+			// different vendors. CDI itself knows better and it will fail the
+			// injection if necessary.
+			log.G(ctx).Warnf("CDI registry refresh failed: %v", err)
+		}
```

Comment:
Correct me if I'm wrong, but it seems GetRegistry returns a singleton object and Refresh takes a lock on the object.

- https://github.com/container-orchestrated-devices/container-device-interface/blob/46367ec063fda9da931d050b308ccd768e824364/pkg/cdi/registry.go
- https://github.com/container-orchestrated-devices/container-device-interface/blob/ee12044b3895afdff3841fee894e3edf01d7fc26/pkg/cdi/cache.go

Wouldn't it introduce a lock contention here? Do we need to refresh the registry every time?

Reply:
> Correct me if I'm wrong, but it seems GetRegistry returns a singleton object and Refresh takes a lock on the object.
> ...
> Wouldn't it introduce a lock contention here? Do we need to refresh the registry every time?

Yes, it is a singleton object which is locked during both Refresh() and CDI device resolution. We don't need to refresh the registry (at least not the brute force way we do now), as long as we make sure in some way that it's up to date. We haven't added any logic for that yet, but we'll get down to it now.

Reply:
@kzys Once we have the implementation for automatic refresh of the CDI registry merged, we won't need to explicitly trigger a manual refresh here.